### PR TITLE
#167313398 Users should receive descriptive signup/Registration validation errors

### PR DESCRIPTION
**Pull request by nkalyesubula**
#### What does this PR do?
 

-  It adds signup validations

#### Description of Task to be completed?

- Users should receive a descriptive error message when they enter invalid user parameters in an attempt to signup

#### How should this be manually tested?

- clone the repo and run **npm run dev**. Then run **http://localhost:8000/api/v1/api-docs/#/default/post_users_signup** in the browser.

#### What are the relevant pivotal tracker stories?

- [#167313398](https://www.pivotaltracker.com/story/show/167313398)

#### Screenshots (if appropriate)

image
image
image
image
image

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ import express from 'express';
 
 import logging from './helpers/logging';
 import routes from './routes/index';
+// import routes from './routes/api/user/user.route';
```
Comment (Mnickii) [inline src/app.js]: remove comment

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "editor.trimAutoWhitespace": false
+}
```
Comment (Mnickii) [inline .vscode/settings.json]: gitignore this file

File: src/controllers/user.controller.js [hunk 3]
```diff
@@ -130,9 +130,9 @@ class UserController {
     try {
       const theUser = await UserService.findOne(req.body.email, '');
       if (theUser) {
-        return res.status(404).send({
-          status: 404,
-          message: `Cannot register User with the id ${req.body.email} which is already in use`
+        return res.status(400).send({
+          status: 400,
```
Comment (Mnickii) [inline src/controllers/user.controller.js]: A better status code for a conflict is 409

**Review by Mnickii — CHANGES_REQUESTED** [review]
Left some comments. Also, please include postman screenshots of the validations.

File: src/controllers/user.controller.js [hunk 4]
```diff
@@ -153,7 +153,7 @@ class UserController {
       const token = await Helper.generateToken(payload);
       return res.status(201).json({
         status: 201,
-        message: 'successfully created account ',
+        message: 'Your account has been successfully created. An email has been sent to you with detailed instructions on how to activate it.',
         data: { firstname, lastname, email },
         token
```
Comment (minega25) [inline src/controllers/user.controller.js:158]: We can set the token in response headers as well

File: src/helpers/helper.js [hunk 5]
```diff
@@ -54,5 +54,16 @@ class Helper {
...
   }
+
+  /**
+   * checks for the existence of any data in the database
+   * @param {object} model The database model.
+   * @param {object} searchParam The search parameter needed to query the database.
+   * @returns {boolean} existing
+   */
+  static async findRecord(model, searchParam) {
+    const existing = await model.findOne({ where: searchParam });
```
Comment (minega25) [inline src/helpers/helper.js]: Where was this method used?

**Review by minega25 — CHANGES_REQUESTED** [review]
Great job on this :-). My review comments are included.

File: src/controllers/user.controller.js (hunk 4 above)
Comment (Mnickii) [reply src/controllers/user.controller.js:158]: Setting tokens in request headers is the recommended way

File: src/middlewares/validators/signup.validation.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      return util.send(res);
+    case '"email" is required':
+      util.setError(400, 'No email was specified');
+      return util.send(res);
+    case '"firstname" is not allowed to be empty':
+      util.setError(400, 'No firstname was specified');
+      return util.send(res);
+    case '"firstname" is required':
+      util.setError(400, 'No firstname was specified');
+      return util.send(res);
```
Comment (Anguandia) [inline src/middlewares/validators/signup.validation.js:73]: I would like to get the difference between case 'first name' is not allowed to be empty and 'first name' is required

File: src/middlewares/validators/signup.validation.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    case '"lastname" is required':
+      util.setError(400, 'No lastname was specified');
+      return util.send(res);
+    case `"lastname" with value "${String(lastname)}" fails to match the required pattern: /^[a-zA-Z]+$/`:
+      util.setError(400, 'lastname should contain only characters');
+      return util.send(res);
+    case '"username" is not allowed to be empty':
+      util.setError(400, 'No username was specified');
+      return util.send(res);
+    case '"username" is required':
```
Comment (Anguandia) [inline src/middlewares/validators/signup.validation.js:89]: Should we make the username a mandatory user input?

**Review by Anguandia — APPROVED** [review]
Good work, edge cases considered as exhaustively as possible except the few lines flagged for clarification ie the difference between required and cannot be empty and if username should really be a mandatory user input

Additionally, the commit message states visiting http://localhost:8000...... in the browser after starting the server; please verify the port information

File: src/middlewares/validators/signup.validation.js (hunk 6 above)
Comment (nkalyesubula) [reply src/middlewares/validators/signup.validation.js:73]: If the key firstname is missing in the request.body, joi error message will be firstname is required but if the key is there but is empty then joi error message will be firstname is not allowed to be empty.  I customized those JOI error messages to descriptive and polite error messages

File: src/middlewares/validators/signup.validation.js (hunk 7 above)
Comment (nkalyesubula) [reply src/middlewares/validators/signup.validation.js:89]: Yes,

**Comment by nkalyesubula** [comment]
> Good work, edge cases considered as exhaustively as possible except the few lines flagged for clarification ie the difference between required and cannot be empty and if username should really be a mandatory user input
> 
> Additionally, the commit message states visiting http://localhost:8000...... in the browser after starting the server; please verify the port information

remember to add PORT=8000 in your .env file. Or use your local server associated port

File: src/helpers/helper.js (hunk 5 above)
Comment (nkalyesubula) [reply src/helpers/helper.js]: Fixed, please check again

File: src/controllers/user.controller.js (hunk 4 above)
Comment (nkalyesubula) [reply src/controllers/user.controller.js:158]: Fixed, please check again

File: src/controllers/user.controller.js (hunk 3 above)
Comment (nkalyesubula) [reply src/controllers/user.controller.js]: Fixed, please check again

File: .vscode/settings.json (hunk 2 above)
Comment (nkalyesubula) [reply .vscode/settings.json]: Fixed, please check again

File: src/app.js (hunk 1 above)
Comment (nkalyesubula) [reply src/app.js]: Fixed, please check again

**Comment by nkalyesubula** [comment]
> Left some comments. Also, please include postman screenshots of the validations.

Added

**Review by Mnickii — APPROVED** [review]
LGTM

File: src/controllers/user.controller.js [hunk 8]
```diff
@@ -153,7 +153,7 @@ class UserController {
       const token = await Helper.generateToken(payload);
       return res.status(201).json({
         status: 201,
-        message: 'successfully created account ',
+        message: 'Your account has been successfully created. An email has been sent to you with detailed instructions on how to activate it.',
```
Comment (nshutijonathan) [inline src/controllers/user.controller.js:156]: The message seems to be very long,I suggest "Your account has been successfully created, activation mail sent"

**Review by nshutijonathan — APPROVED** [review]

**Review by salviosage — APPROVED** [review]
LGTM

**Pull request merged**